uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/LeaveRequest.php

```diff
@@ -920,28 +920,31 @@ public static function findOverlappingLeaveRequests($leaveRequestParams, $leaveR
       ? "lr.from_date < '{$toDateTime}' AND lr.to_date > '{$fromDateTime}'"
       : "lrd.date BETWEEN '{$fromDate}' AND '{$toDate}'";
 
-    if ($isCalculationUnitInHours) {
-      $absenceTypesSelector = 'AND lr.type_id IN (' . implode(',', $absenceTypesIDsWithSameUnit) . ')';
-
-      if (count($absenceTypesIDsWithDifferentUnit)) {
-        $considerRequestsInDifferentUnit =
-          "OR ((lrd.date BETWEEN '{$fromDate}' AND '{$toDate}') AND lr.type_id IN (" .
-          implode(',', $absenceTypesIDsWithDifferentUnit) . "))";
+    if (!$isTOIL) {
+      if ($isCalculationUnitInHours) {
+        $absenceTypesSelector = 'AND lr.type_id IN (' . implode(',', $absenceTypesIDsWithSameUnit) . ')';
+
+        if (count($absenceTypesIDsWithDifferentUnit)) {
+          $considerRequestsInDifferentUnit =
+            "OR ((lrd.date BETWEEN '{$fromDate}' AND '{$toDate}') AND lr.type_id IN (" .
+            implode(',', $absenceTypesIDsWithDifferentUnit) . "))";
+        }
+      } else {
+        $absenceTypesSelector = 'AND lr.type_id IN (' . implode(',', $absenceTypesIDs) . ')';
       }
     }
 
-    if (!$isCalculationUnitInHours && !$isTOIL) {
-      $absenceTypesSelector = 'AND lr.type_id IN (' . implode(',', $absenceTypesIDs) . ')';
-    }
-
+    $ignoreRequestsWithDifferentUnit = count($absenceTypesIDsWithDifferentUnit)
+      ? 'OR lr.type_id IN (' . implode(',', $absenceTypesIDsWithDifferentUnit) . ')'
+      : '';
     $considerHalfDayRequests =
       ($fromDateType === $halfDayPMID
         ? "AND (DATE_FORMAT(lr.to_date, '%Y-%m-%d') != '{$fromDate}' " .
-          "OR lr.to_date_type != {$halfDayAMID})"
+          "OR lr.to_date_type != {$halfDayAMID} {$ignoreRequestsWithDifferentUnit})"
```

Comment:
This query is not easy to understand, would have been easier to add the condition for NOT over lapping and negate it just like I suggested in the first PR review.

Reply:
Attempted to improve.